### ISSUE TITLE
CPU hogging issue with Linux

### ISSUE THREAD

**Issue by samdivaio** [Kwoth/NadekoBot#294]
**Discord is hogging cpu whenever bot's voice activity is stopped**

`mono NadekoBot.exe` uses 2-10% of cpu on regular activities,
while playing music it extends to 10-20% (extreme case)
however, when the **music is paused, finished or stopped** `(any possible time while bot is in voice channel and not playing/doing anything)` it uses 90-99% of cpu `(a full one core if the machine have more than one core cpu, so if the machine has two cores, then it hogs 1st core most of the time and occasionally leaves 1st core and hogs the second core so it hogs one core anyway)`

The issue it creates is (if the bot is left that way, for awhile) it doesn't respond to music commands on that server, meaning sometimes the `!m q ` works even if bot doesn't respond or reply anything and sometimes it makes `!m d` and `!m cleanup` unusable. and only `.restart` can fix it, and sometime the whole music module is inoperable while everything else works.

Possible fix to this would be making bot auto leave voice channel every time bot stops playing music, like function of `!m d` implemented when bot finishes playing music, when `!m s` command is used and when `!m p` command is used as well.
Or a possible workaround is making bot auto play a blank sound/song in loop when no song is played or bot is paused to keep the bots "voice activity" in voice channel active.

**Comment by coggy9** [comment]
I can confirm this. [The CPU graph on my VPS ](http://i.imgur.com/5TQJMtR.png) shoots up a few minutes after music stops playing.

**Comment by Kwoth** [comment]
Can someone confirm this happening on the latest release?

**Comment by samdivaio** [comment]
Unfortunately it's still happening! Probably nothing to do with bot but maybe discord libraries

**Comment by aylamar** [comment]
I have the same issue running NadekoBot v0.98 on Ubuntu 14.04.4 x64.

**Comment by Lemmmy** [comment]
can confirm

**Comment by Lemmmy** [comment]
the issue is potentially related to how mono handles async, and this line: https://github.com/Kwoth/NadekoBot/blob/master/NadekoBot/Modules/Music/Classes/MusicControls.cs#L67

**Comment by Kwoth** [comment]
Ok I'm starting work on this right now.

**Comment by ihatevim** [comment]
I saw the new buffer changes, I'm not sure whether they were relevant to this bug but it's still happening. Once a song finishes playing, CPU usage spikes to 100% breaking the module entirely.

**Comment by ihatevim** [comment]
Okay so forgive me if this is out of the ballpark since I have little to no knowledge of C# but
`                        }
                        finally
                        {
                            await Task.Delay(300).ConfigureAwait(false);
CurrentSong = null;`

Doesn't this mean that when a song ends it's permanently in the while(!destroyed) loop? Shouldn't you completely exit the loop once a song is finished. If it's just iterating through dead if statements that could be why it's hogging CPU.

**Comment by Kwoth** [comment]
Absolutely no idea why does it not work properly on mono. Tried on several windows servers/PCs and all have 0-2% cpu usage per song. On mono with the newest dev is still at 99% the whole time.

**Comment by Lemmmy** [comment]
solution: remove c#

**Comment by frostyfrog** [comment]
@Lemmmy mentioned that it could be related to how mono uses async, however, it appears that the other async that is spun off in it's own thread works just fine. I'm curious if the code is somehow getting stuck in an infinite loop instead and isn't waiting on the 100ms delay.

I'm not sure, I'm not experienced with C# and I'm just throwing things at the wall. I just happened to notice this after Linode sent me an email about high CPU usage for more than 2 hours. :worried: If I could be of any assistance in any way, just let me know. I'm currently using the docker container build since my host is using ArchLinux and has the latest and "greatest" system updates.

**Comment by Nitix** [comment]
The 1.0 release will run on dotnet, that is natively crossplatform so this bug shouldn't appear with this release.

**Comment by Kwoth** [comment]
closing as this wont happen on 1.0

**Comment by AterialDawn** [comment]
I actually found the cause of this bug, its a problem with Task.Delay in Discord.Net.Audio, specifically line 387 of VoiceSocket.cs
```CSharp
await Task.Delay(1).ConfigureAwait(false); //Give as much time to the encrypter as possible
```
Mono seems to bug out when Task.Delay(1) is used and locks up the cpu when it should be sleeping. Just changing this line to 
```CSharp
await Task.Delay(2).ConfigureAwait(false);
```
fixes it for me.  CPU usage when idling is now 35% (still pretty high but at least it doesn't peg my CPU to 100%). Setting the delay to 10ms drops the idle usage from 35% to 10% but introduces occasional stuttering when playing back music.

**Comment by Lemmmy** [comment]
@AterialDawn make a PR?

**Comment by Nitix** [comment]
Well, it isn't a bug with discord.net but with Mono. Also, we won't do another update for Nadeko 0.9.x. 1.0 will come soon, I don't know when but Nadeko public has 1.0 and Kwoth is currently fixing bugs.

Also this "fix" isn't a correct way to fix it as it introduce occasional stuttering.

I can confirm that this bug isn't in Nadeko 1.0. I tested it with Ubuntu 16.04.

**Comment by AterialDawn** [comment]
I've been testing it for the last day or so on 3 different servers at the same time, with the delay set to 3 there is no noticeable stuttering, and cpu usage is at a very acceptable 15% on idle. I haven't made a pull request because, as Nitix has said, it's a bug with mono and this is only a bypass to deal with that mono issue.
